exchanges/kucoin/kucoin_websocket.go

```diff
@@ -913,65 +887,51 @@ func (ku *Kucoin) processOrderbook(respData []byte, symbol string) error {
 		return err
 	}
 	var init bool
-	assetEnabledPairs := ku.listOfAssetsCurrencyPairEnabledFor(pair)
-	if assetEnabledPairs[asset.Spot] && ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
-		init, err = ku.UpdateLocalBuffer(response, asset.Spot)
-		if err != nil {
-			if init {
-				return nil
-			}
-			return fmt.Errorf("%v - UpdateLocalCache for asset type %v error: %s",
-				ku.Name,
-				asset.Spot,
-				err)
+	for _, assetType := range ku.listOfAssetsCurrencyPairEnabledFor(pair) {
+		if ku.CurrencyPairs.IsAssetEnabled(assetType) != nil {
+			continue
 		}
-	}
-	if assetEnabledPairs[asset.Margin] && ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
-		init, err = ku.UpdateLocalBuffer(response, asset.Margin)
+		init, err = ku.UpdateLocalBuffer(response, assetType)
 		if err != nil {
 			if init {
 				return nil
 			}
 			return fmt.Errorf("%v - UpdateLocalCache for asset type %v error: %s",
 				ku.Name,
-				asset.Margin,
+				assetType,
 				err)
 		}
 	}
 	return nil
 }
 
-func (ku *Kucoin) processMarketSnapshot(respData []byte, instrument string) error {
+func (ku *Kucoin) processMarketSnapshot(respData []byte) error {
 	response := WsSpotTicker{}
 	err := json.Unmarshal(respData, &response)
 	if err != nil {
 		return err
 	}
-	pair, err := currency.NewPairFromString(instrument)
+	pair, err := currency.NewPairFromString(response.Data.Symbol)
 	if err != nil {
 		return err
 	}
-	spotTickerPrice := ticker.Price{
-		ExchangeName: ku.Name,
-		AssetType:    asset.Spot,
-		Last:         response.Data.LastTradedPrice,
-		Pair:         pair,
-		Low:          response.Data.Low,
-		High:         response.Data.High,
-		QuoteVolume:  response.Data.VolValue,
-		Volume:       response.Data.Vol,
-		Open:         response.Data.Open,
-		Close:        response.Data.Close,
-		LastUpdated:  response.Data.Datetime.Time(),
-	}
-	assetEnabledPairs := ku.listOfAssetsCurrencyPairEnabledFor(pair)
-	if assetEnabledPairs[asset.Spot] && ku.AssetWebsocketSupport.IsAssetWebsocketSupported(asset.Spot) && ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
-		ku.Websocket.DataHandler <- &spotTickerPrice
-	}
-	if assetEnabledPairs[asset.Margin] && ku.AssetWebsocketSupport.IsAssetWebsocketSupported(asset.Margin) && ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
-		marginTickerPrice := spotTickerPrice
-		marginTickerPrice.AssetType = asset.Margin
-		ku.Websocket.DataHandler <- &marginTickerPrice
+	for _, assetType := range ku.listOfAssetsCurrencyPairEnabledFor(pair) {
+		if !ku.AssetWebsocketSupport.IsAssetWebsocketSupported(assetType) && ku.CurrencyPairs.IsAssetEnabled(assetType) != nil {
```

Comment:
Both these can be removed I think:
* AssetWebsocketSupport is marked as deprecated and I don't think we need to respect it *either* for Kucoin in general, or in the context of having received an update
* ku.CurrencyPairs.IsAssetEnabled is redundant. We got the assets from GetAssetTypes(enabled) AND GetEnabledPairs(a) also checks that the asset is enabled.
So I think you can remove these two conditions and the continue without risk.

Reply:
removed